File: Java/src/com/leetcode/Medium/BinaryTreeZigzagLevelOrderTraversal.java
package com.leetcode.Medium;

import java.util.*;

// 103. Binary Tree Zigzag Level Order Traversal
public class BinaryTreeZigzagLevelOrderTraversal {
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null)
            return result;

        boolean leftToRight = true;
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        while (!queue.isEmpty()) {
            int size = queue.size();
            List<Integer> list = new ArrayList<>();
            for (int i = 0; i < size; i++) {
                TreeNode node = queue.poll();

                if (leftToRight)
                    list.add(node.val);
                else
                    list.add(0, node.val);

                if (node.left != null)
                    queue.offer(node.left);
                if (node.right != null)
                    queue.offer(node.right);
            }
            leftToRight = !leftToRight;
            result.add(list);
        }

        return result;
    }
}

File: JavaScript/Medium/BinaryTreeZigzagLevelOrderTraversal.js
/**
 * 103. Binary Tree Zigzag Level Order Traversal
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function (root) {
  let result = [];
  if (!root) return result;

  let leftToRight = true;
  let queue = [];
  queue.push(root);
  while (queue.length > 0) {
    let size = queue.length;
    let list = [];
    for (let i = 0; i < size; i++) {
      let node = queue.shift();

      if (leftToRight) {
        list.push(node.val);
      } else {
        list.unshift(node.val);
      }

      if (node.left) {
        queue.push(node.left);
      }
      if (node.right) {
        queue.push(node.right);
      }
    }
    leftToRight = !leftToRight;
    result.push(list);
  }

  return result;
};
